hudi-flink-datasource/hudi-flink/src/main/java/org/apache/hudi/table/format/mor/MergeOnReadInputFormat.java

```diff
@@ -166,29 +177,56 @@ public void open(MergeOnReadInputSplit split) throws IOException {
     this.currentReadCount = 0L;
     this.closed = false;
     this.hadoopConf = HadoopConfigurations.getHadoopConf(this.conf);
+    if (schemaEvolutionContext.isPresent()) {
+      SchemaEvolutionContext context = schemaEvolutionContext.get();
+      querySchema = context.getQuerySchema();
+      actualSchema = context.getActualSchema(split);
+      actualFieldNames = context.getFieldNames(actualSchema);
+      actualFieldTypes = context.getFieldTypes(actualSchema);
+    } else {
+      querySchema = InternalSchema.getEmptyInternalSchema();
+      actualSchema = InternalSchema.getEmptyInternalSchema();
+      actualFieldNames = fieldNames;
+      actualFieldTypes = fieldTypes;
+    }
+
     if (!(split.getLogPaths().isPresent() && split.getLogPaths().get().size() > 0)) {
-      if (split.getInstantRange() != null) {
+      if (split.getInstantRange().isPresent()) {
         // base file only with commit time filtering
         this.iterator = new BaseFileOnlyFilteringIterator(
-            split.getInstantRange(),
-            this.tableState.getRequiredRowType(),
+            split.getInstantRange().get(),
             getReader(split.getBasePath().get(), getRequiredPosWithCommitTime(this.requiredPos)));
+        int[] positions = IntStream.range(1, requiredPos.length + 1).toArray();
+        RowDataProjection projection = getCastProjection(positions)
+            .orElse(RowDataProjection.instance(tableState.getRequiredRowType(), positions));
+        projectRecordIterator(projection);
       } else {
         // base file only
         this.iterator = new BaseFileOnlyIterator(getRequiredSchemaReader(split.getBasePath().get()));
+        projectRecordIterator();
       }
     } else if (!split.getBasePath().isPresent()) {
       // log files only
       if (OptionsResolver.emitChangelog(conf)) {
         this.iterator = new LogFileOnlyIterator(getUnMergedLogFileIterator(split));
+        projectRecordIterator();
       } else {
         this.iterator = new LogFileOnlyIterator(getLogFileIterator(split));
+        projectRecordIterator();
```

Comment:
No need to call projectRecordIterator() as the record has been rewriten in  https://github.com/apache/hudi/blob/master/hudi-common/src/main/java/org/apache/hudi/common/table/log/AbstractHoodieLogRecordReader.java#L389